xarray/coding/variables.py

```diff
@@ -520,9 +520,19 @@ def encode(self, variable: Variable, name: T_Name = None) -> Variable:
             # trying to get it from encoding, resort to an int with the same precision as data.dtype if not available
             signed_dtype = np.dtype(encoding.get("dtype", f"i{data.dtype.itemsize}"))
             if "_FillValue" in attrs:
-                new_fill = np.array(attrs["_FillValue"])
-                # use view here to prevent OverflowError
-                attrs["_FillValue"] = new_fill.view(signed_dtype).item()
+                try:
+                    # user provided the on-disk signed fill
+                    new_fill = signed_dtype.type(attrs["_FillValue"])
+                except OverflowError:
```

Comment:
Thinking about it now, the `except` clause is probably the common case as far as xarray is concerned. That is, data loaded with xarray will use the decode logic which will use the unsigned integer version of the fill value. However, for anyone constructing their own DataArray/Dataset objects, there is a good chance of providing the signed version of the fill value (the `try` clause).

Reply:
And I should have said, my original thinking in the related issue was that `_FillValue` should always be the on-disk type, but then realized it makes sense to have it match the in-memory loaded data type. I too would like there to be only one way of doing it, but I'm not sure how likely that is. I suppose the current `except` clause should be the preferred way?

Reply:
The decode logic will treat `-1` and `255` the same for on-disk `int8` type (as intended).

> my original thinking in the related issue was that _FillValue should always be the on-disk type, but then realized it makes sense to have it match the in-memory loaded data type.

CF says _FillValue must be the on-disk type.

Reply:
If I understood @kmuehlbauer, the in-memory type is needed for masking decoding to work, but also said that maybe the unsigned and masking should be combined. I could also see an argument for it being wrong to have a `_FillValue` that doesn't work for and match the data it is attached to (in-memory).

Reply:
If we/you think that it is OK for in-memory `_FillValue` to be the on-disk fill all the time even when it doesn't match the in-memory array type, then I can look at combining the two Coders (`UnsignedIntegerCoder` and `CFMaskCoder`).

Reply:
At decode time,
https://github.com/pydata/xarray/blob/9166eb2cc7ad4b57983396fb0632e7b0db15e278/xarray/coding/variables.py#L538-L545

the written fill value is read as signed type and cast to unsigned type and saved in `_FillValue` which is then used next for masking

https://github.com/pydata/xarray/blob/9166eb2cc7ad4b57983396fb0632e7b0db15e278/xarray/conventions.py#L280-L286

Reply:
> then I can look at combining the two Coders (UnsignedIntegerCoder and CFMaskCoder

I don't think this is necessary at the moment because the decode pipeline overwrites the `_FillValue` after casting to unsigned.

Reply:
AFAICT the error here was that numpy will chose 64-bit data types for python integers, and those cannot be "viewed" as lower-precision dtypes. And our solution is to be clear about what dtype is specified which CF claims should always be the on-disk dtype.

Reply:
That's correct for the encoding pipeline. I can add a serialization warning for the case where the provided `_FillValue` is not already the on-disk signed type.

For the decoding, you say that it already overwrites the `_FillValue` after casting to unsigned. Which part overwrites `_FillValue`? The `CFMaskCoder` uses the unsigned `_FillValue` and retains it in `.encoding["_FillValue"]` from my tests with a real NetCDF file. I'm also just realizing that xarray `.open_dataset` always converts a pure integer variable to a floating point variable and replaces fills with NaN.

Reply:
> I can add a serialization warning for the case where the provided _FillValue is not already the on-disk signed type.

Won't this warn for python integers though?

> The CFMaskCoder uses the unsigned _FillValue and retains it in .encoding["_FillValue"] from my tests with a real NetCDF file.

Ah yes this is correct, we don't reset the type after masking and that would be why we want to combine UnsignedIntegerCoder and CFMaskCoder.

I think we can combine these in a future PR

>  I'm also just realizing that xarray .open_dataset always converts a pure integer variable to a floating point variable and replaces fills with NaN.

Yes.

Reply:
>>  I can add a serialization warning for the case where the provided _FillValue is not already the on-disk signed type.
>
> Won't this warn for python integers though?

To be clear this would be in the encode pipeline and would only happen if the fill provided doesn't fit into the on-disk dtype.

```python
In [9]: np.dtype(np.int8).type(255)
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[9], line 1
----> 1 np.dtype(np.int8).type(255)

OverflowError: Python integer 255 out of bounds for int8

In [10]: np.dtype(np.int8).type(-1)
Out[10]: np.int8(-1)

In [11]: np.__version__
Out[11]: '2.0.0'

```

So 255 -> int8 is error, -1 -> int8 is fine.

BUT this means anyone doing a `xr.open_dataset` followed by a `.to_netcdf` will get the serialization warning with the logic as is (`main` and this PR) because the `_FillValue` was made unsigned in the decode pipeline.

Reply:
> BUT this means anyone doing a xr.open_dataset followed by a .to_netcdf will get the serialization warning with the logic as is (main and this PR) because the _FillValue was made unsigned in the decode pipeline.

OK good argument.

It *seems* to me like this is good to go then? And we can follow up with a PR that merges UnsignedCoder and MaskCoder, and handles the FillValue properly. Does that sound right to you?

Reply:
Without the new serialization warning, yes, I think this is as I expect. Right now this PR should at least restore behavior. If anything, especially based on this discussion, this code is too flexible...but that's kind of how anyone claiming they make CF-compliant NetCDF files assumes things will work.